sql/core/src/main/java/org/apache/spark/sql/sources/v2/WriteMicroBatchSupport.java

```diff
@@ -0,0 +1,40 @@
+package org.apache.spark.sql.sources.v2;
+
+import java.util.Optional;
+
+import org.apache.spark.annotation.InterfaceStability;
+import org.apache.spark.sql.sources.v2.writer.DataSourceV2Writer;
+import org.apache.spark.sql.streaming.OutputMode;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * A mix-in interface for {@link DataSourceV2}. Data sources can implement this interface to
+ * provide data writing ability and save the data from a microbatch to the data source.
+ */
+@InterfaceStability.Evolving
+public interface WriteMicroBatchSupport extends BaseStreamingSink {
+
+  /**
+   * Creates an optional {@link DataSourceV2Writer} to save the data to this data source. Data
+   * sources can return None if there is no writing needed to be done.
+   *
+   * If this method fails (by throwing an exception), the action would fail and no Spark job was
+   * submitted.
+   *
+   * @param queryId A unique string for the writing query. It's possible that there are many writing
+   *                queries running at the same time, and the returned {@link DataSourceV2Writer}
+   *                can use this id to distinguish itself from others.
```

Comment:
By the same logic, even the createReader() should take queryId as a parameter to allow concurrent readers to distinguish among themselves. This is right way to distinguish between concurrent cached KafkaConsumers (as opposed to our current way of using generated groupids).

Reply:
I think the scenarios aren't quite parallel. It's okay for readers to not acknowledge the continuity between restarts of the same query; as long as they're passed in the appropriate offsets, they'll behave indistinguishably. But some sinks may want to know whether a writer is trying to re-commit part of a query that's already committed.

The core problem is that we can't pass queryId to createReader() in the batch v2 model, because it's invoked in the DataFrameReader before there's a query to ID. We could pass it in the streaming case, since we don't create the reader until {Stream/Continuous}Execution, but that would require an additional API divergence.